src/EditorFeatures/CSharpTest/CodeActions/ConvertLinq/ConvertForEachToLinqQueryTests.cs

```diff
@@ -257,7 +455,21 @@ IEnumerable<int> M()
                select x;
     }
 }";
-            await TestInRegularAndScriptAsync(source, output);
+            await TestInRegularAndScriptAsync(source, queryOutput, index: 0);
+
+            string linqInvocationOutput = @"
+using System;
+using System.Linq;
+class C
+{
+    IEnumerable<int> M()
+    {
+        var nums = new int[] { 1, 2, 3, 4 };
+        return nums.Where(x => x > 2).Select(x => x);
```

Comment:
i have a small concern here.  Namely that this can lead to uber-long invocation chains.  One thing i think we should do is something similar to what i did with "convert to conditional".  Namely, that if the line gets too long (i.e. past 120 columns) that we smartly wrap.  so use-conditional wraps to:

```c#
long_condition
    ? long_true_part
    : long_false_part;
```

We also do this for things like 'generate-equals'.  we don't want to make an excessively long line, so we generate:

```c#
return this.x == other.x &&
       this.y == other.y &&
       ...
```

I think we should do that here as well.  once it gets too long we wrap (and align on invocations), making something like:

```c#
return nums.Where(x => x > 2)
           .Select(x => x);
```

It will make the system feel a lot better and less mechanical.

Reply:
I agree that something needs to be done here to make the output better for such cases. However, I feel this should not be the job of each code fix/refactoring provider, but instead should be part of a post-process code cleanup pass, which runs the formatting analyzer and bunch of other user configured cleanup to ensure the newly generated code matches their code style preferences (wrap threshold in terms of column count, wrap at which syntax node for nested lambda invocations (start of invocation, argument list or lambda body), expression bodied vs block bodied lambdas, etc. Tagging @heejaechang @sharwell - does formatting analyzer/fixer handle such preferences?

Reply:
Currently, formatting doesn't have anything baked in for this. 

> which runs the formatting analyzer and bunch of other user configured cleanup to ensure the newly generated code matches their code style preferences 

The issue here is that there tends to be two types of preferences.  Preferences for code you have written, and have decided on how it should look, and preferences for code that is totally new.  

So, for example, if i did 'extract method' on a bunch of statements, i would *not* want those statements to have their formatting touched (except to maybe change indentation).  However, if i'm effectively generating a large swath of new code+constructs, it would be good to have that code be moderately 'clean', even for areas we have no concept of formattin for (like wrapping).

While it would be nice to push this off to formatting (and i'm fine with teh code living there), the truth is that this work is unlikely to happen unless people who are makign the features go create that system as well.  Otherwise, it simply will not happen.

Reply:
@heejaechang it's probably worth investing in some sort of "smarter" wrapping system in the formatter, even if it's just for code with elastic annotations.  Basically, since we know the code is new/fresh (thanks to the annotations), we should be a bit more aggressive avoiding poor looking code (i.e. code that is exceptionally long).  

We could have an option, for example, like:  PreferredWrappingColumnForElasticCode.  This could then be associated with targeted high-value formatting tweaks for certain cases.  The cases i could think off right off the top of my head would be:

1. parameter list wrapping when doing an extract/generate method.  Right now it *is* annoying that those can generate parameter lists that are several hundred chars long.
2. arg list wrapping for extract-method.  
3. wrapping of long binary expressions.
4. wrapping of long invocation expressions.

It feels like we could make some targeted fixes here, and only for new/elastic code.  The major downside i could see woudl be around our wrapping not fitting a particular style the user wants.  For example, for invocation wrapping, some may want:

```c#
return nums.Where(x => x > 2)
           .Select(x => x);

// vs

return nums.Where(x => x > 2).
            Select(x => x);
```

For parameter wrapping, some will want:

```c#
void Foo(int a, int b, int c,
         int d, int e, int f);

// vs

void Foo(int a, int b, int c,
    int d, int e, int f);
```

There's risk of the cure being worse than the disease if people then are more put off by our formatting decisions not being what they want, vs us not formatting at all.

Reply:
I think this fits well into the system that @heejaechang and @ivanbasov have talked about in the past and heejae filed https://github.com/dotnet/roslyn/issues/25844 - have a separate trivia handling service that code fixers/refactorings can just register with it on the input syntax at start of analysis, and request it to process the final output syntax to automatically add back any comments from dropped nodes/tokens and also format it appropriately to add necessary newline, whitespace and indentation.

Reply:
Perhaps :)   though i would prefer they be handled separately.  If only because i'm uncertain about the feasiability of the trivia-preserving facility.  While, on the other hand, just doing a better job of formatting def seems feasible and much lower cost.